In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# create shortcut path 
path = r"C:\Users\HP\Documents\06-2024 Instacart Basket Analysis" 

In [3]:
# import data set
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
df_cust.shape

(206209, 10)

# 04. Data Wrangling

In [7]:
# renaming columns
df_cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)
df_cust.rename(columns = {'Surnam' : 'last_name'}, inplace = True)
df_cust.rename(columns = {'Gender' : 'gender'}, inplace = True)
df_cust.rename(columns = {'STATE' : 'state'}, inplace = True)
df_cust.rename(columns = {'Age' : 'age'}, inplace = True)

In [8]:
df_cust.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [16]:
# finding data types
df_cust.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [18]:
# changing data types
df_cust['user_id'] = df_cust['user_id'].astype('int32')
df_cust['age'] = df_cust['age'].astype('int32')
df_cust['n_dependants'] = df_cust['n_dependants'].astype('int32')
df_cust['income'] = df_cust['income'].astype('int32')

In [20]:
# data type check
df_cust.dtypes

user_id          int32
first_name      object
last_name       object
gender          object
state           object
age              int32
date_joined     object
n_dependants     int32
fam_status      object
income           int32
dtype: object

I elected to change the 'int64' data types to 'int32' data types to decrease memory usage.

# 05. Data Quality and Consistency Checks

In [23]:
# finding missing values
df_cust.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

There are 11,259 missing values in the dataframe. The missing field will not be a detriment to the analysis so I am electing to keep them as is.

In [27]:
# checking for mixed-data types
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].map(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)
else : print ("no mixed-type data")

first_name
no mixed-type data


In [29]:
# fixing "first_name" column for mixed-data
df_cust['first_name'] = df_cust['first_name'].astype('str')

In [31]:
# verifying for mixed-data types
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].map(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)
else : print ("no mixed-type data")

no mixed-type data


In [33]:
# finding duplicates
df_dups = df_cust[df_cust.duplicated()]

In [35]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

In [44]:
df_dups.shape

(0, 10)

In [37]:
num_duplicates = df_dups.sum()

In [39]:
num_duplicates

user_id         0
first_name      0
last_name       0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: object

There are no duplicates in this dataframe. The current df (df_cust) is good as is.

In [42]:
# consistency checks
df_cust.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [50]:
date_min = df_cust['date_joined'].min()

In [52]:
date_max = df_cust['date_joined'].max()

In [56]:
date_min

'1/1/2017'

In [58]:
date_max

'9/9/2019'

There seems to be no discrepancies in the consistency checks. The only modifications were name changes, and data type changes

In [64]:
# final review
df_cust.shape

(206209, 10)

In [66]:
df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [70]:
df_cust.tail()

user_id first_name last_name  gender           state  age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

# 06. Merging datasets

In [61]:
# import 'ords_prods_merge'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_aggregated.pkl'))

In [72]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Regularly busy           Most orders   
1  Mid-range product  Regularly busy   Regularly busy        Average orders   
2  Mid-range product     Busiest day     Busiest days        Average orders   
3  Mid-range product  Regularly busy  Least busy days           Most orders   
4  Mid-range product      Least busy  Least busy days        Average orders   

  max_order      loyalty_flag average_spend  spending_flag customer_frequency  \
0        32  Regular customer      6.935811    Low spender                8.0   
1        32  Regular customer      6.935811    Low spender                8.0   
2         5      New customer      7.930208    Low spender                8.0   
3         3      New customer      4.972414    Low spender                9.0   
4         3      New customer      4.972414    Low spender                9.0   

      frequency_flag  
0  Frequent customer  
1  Frequent customer  
2  Frequent customer  
3  Frequent customer  
4  Frequent customer  

[5 rows x 23 columns]

In [74]:
ords_prods_merge.shape

(32404859, 23)

In [76]:
# merging data using "user_id" as the commonality
ords_prods_all = ords_prods_merge.merge(df_cust, on = 'user_id', how = 'inner', indicator = True)

I am using an inner join for the merge because the information from the customer dataframe is only valuable when using with the information from the 'ords_prods_merge' dataframe.

In [78]:
ords_prods_all.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

   first_name  last_name  gender      state age date_joined n_dependants  \
0     Charles        Cox    Male  Minnesota  81    8/1/2019            1   
1     Charles        Cox    Male  Minnesota  81    8/1/2019            1   
2     Deborah      Glass  Female    Vermont  66   6/16/2018            2   
3     Heather      Myers  Female  Wisconsin  40    2/9/2020            3   
4     Heather      Myers  Female  Wisconsin  40    2/9/2020            3   

   fam_status  income  _merge  
0     married   49620    both  
1     married   49620    both  
2     married  158302    both  
3     married   31308    both  
4     married   31308    both  

[5 rows x 33 columns]

In [80]:
ords_prods_all['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [82]:
# removing '_merge' column
ords_prods_all.drop(columns = "_merge", inplace = True)

In [84]:
ords_prods_all.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

      frequency_flag  first_name  last_name  gender      state age  \
0  Frequent customer     Charles        Cox    Male  Minnesota  81   
1  Frequent customer     Charles        Cox    Male  Minnesota  81   
2  Frequent customer     Deborah      Glass  Female    Vermont  66   
3  Frequent customer     Heather      Myers  Female  Wisconsin  40   
4  Frequent customer     Heather      Myers  Female  Wisconsin  40   

  date_joined  n_dependants fam_status  income  
0    8/1/2019             1    married   49620  
1    8/1/2019             1    married   49620  
2   6/16/2018             2    married  158302  
3    2/9/2020             3    married   31308  
4    2/9/2020             3    married   31308  

[5 rows x 32 columns]

In [86]:
ords_prods_all.shape

(32404859, 32)

In [88]:
ords_prods_all.dtypes

product_id                  int32
product_name               object
aisle_id                    int32
department_id               int32
prices                    float32
order_id                    int32
user_id                     int32
order_number                int32
order_day_of_week           int32
order_hour_of_day           int32
days_since_prior_order    float32
add_to_cart_order           int32
reordered                   int32
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int32
loyalty_flag               object
average_spend             float32
spending_flag              object
customer_frequency        float32
frequency_flag             object
first_name                 object
last_name                  object
gender                     object
state                      object
age                         int32
date_joined                object
n_dependants  

# 08. Exporting new dataframe

In [93]:
ords_prods_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))